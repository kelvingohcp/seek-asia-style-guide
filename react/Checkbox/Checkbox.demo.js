import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Checkbox from './Checkbox';
import * as sketch from './Checkbox.sketch';
import demoStyles from './Checkbox.demo.less';
import classnames from 'classnames';

class CheckboxContainer extends Component {
  static propTypes = {
    component: PropTypes.func.isRequired,
    componentProps: PropTypes.object.isRequired
  };

  constructor() {
    super();

    this.state = {
      checked: false
    };
  }

  handleChange = event => {
    this.setState({
      checked: event.target.checked
    });
  };

  render() {
    const { component: DemoComponent, componentProps } = this.props;
    const { checked } = this.state;
    const classNames = classnames({
      [demoStyles.root]: true, // eslint-disable-line
      [demoStyles.fullWidth]: componentProps.fullWidth
    });

    return (
      <div className={classNames}>
        <DemoComponent
          {...componentProps}
          checked={checked}
          onChange={this.handleChange}
        />
      </div>
    );
  }
}

export default {
  route: '/checkbox',
  title: 'Checkbox',
  component: Checkbox,
  container: CheckboxContainer,
  sketch,
  initialProps: {
    id: 'stillInRole',
    label: 'Still in role',
    extraLabel: '(All)',
    // Documentation only:
    checked: false,
    onChange: () => {}
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Full width',
          transformProps: props => ({
            ...props,
            fullWidth: true
          })
        }
      ]
    },
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Compact',
          transformProps: props => ({
            ...props,
            compact: true
          })
        },
        {
          label: 'Disabled',
          transformProps: props => ({
            ...props,
            inputProps: {
              disabled: true
            }
          })
        }
      ]
    },
    {
      label: 'Checkbox position',
      type: 'radio',
      states: [
        {
          label: 'Left',
          transformProps: props => ({
            ...props
          })
        },
        {
          label: 'Right',
          transformProps: props => ({
            ...props,
            rtl: true
          })
        }
      ]
    }
  ]
};
