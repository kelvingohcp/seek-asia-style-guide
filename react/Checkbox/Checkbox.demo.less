.root > :global(*) {
  min-width: 150px; // :(
  max-width: 350px;
}

.fullWidth {
  width: 300px;
}
