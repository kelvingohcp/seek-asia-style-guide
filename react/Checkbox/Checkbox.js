import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './Checkbox.less';
import classnames from 'classnames';
import { Icon, Text } from 'seek-asia-style-guide/react';

function combineClassNames(props = {}, ...classNames) {
  const { className, ...restProps } = props;

  return {
    className: classnames.apply(null, [...classNames, className]), // eslint-disable-line no-useless-call
    ...restProps
  };
}

export default class Checkbox extends Component {
  static displayName = 'Checkbox';

  static propTypes = {
    id: PropTypes.string.isRequired,
    label: PropTypes.node,
    extraLabel: PropTypes.node,
    value: PropTypes.string,
    className: PropTypes.string,
    checked: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    inputProps: PropTypes.object,
    rtl: PropTypes.bool,
    fullWidth: PropTypes.bool,
    compact: PropTypes.bool,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    checked: false,
    inputProps: {},
    rtl: false,
    fullWidth: false,
    compact: false,
    disabled: false
  };

  state = {
    checked: false,
    disabled: false
  }

  static getDerivedStateFromProps(props) {
    return {
      checked: props.checked || props.inputProps && props.inputProps.checked,
      disabled: props.disabled || props.inputProps && props.inputProps.disabled
    };
  }

  renderInput() {
    const { id, value, checked, disabled, onChange, onFocus, onBlur, inputProps } = this.props;

    const allInputProps = {
      id,
      value,
      checked,
      disabled,
      onChange,
      onFocus,
      onBlur,
      ...combineClassNames(inputProps, styles.input),
      type: 'checkbox'
    };

    return <input {...allInputProps} />;
  }

  render() {
    const { checked, className, compact, disabled, extraLabel, fullWidth, id, label, rtl } = this.props;
    const rootClassNames = classnames({
      [styles.root]: true,
      [className]: className,
      [styles.checked]: checked,
      [styles.compact]: compact,
      [styles.disabled]: disabled,
      [styles.fullWidth]: fullWidth,
      [styles.rtl]: rtl
    });

    return (
      <label className={rootClassNames} htmlFor={id}>
        {this.renderInput()}
        <div className={styles.checkBox}>
          <Icon type="check" svgClassName={styles.checkMark} />
        </div>
        {(label || extraLabel) && (
          <div className={styles.label}>
            <Text baseline={false} intimate={compact} strong={checked} className={styles.labelText}>{label}</Text>
            {extraLabel && <Text baseline={false} whispering strong={checked} className={styles.labelTextSecondary}>{extraLabel}</Text>}
          </div>)}
      </label>
    );
  }
}
