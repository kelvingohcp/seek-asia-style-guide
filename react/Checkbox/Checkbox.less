@import (reference) '~seek-asia-style-guide/theme';

.activeFocusReset() {
  .input:not([readonly]):not([disabled]):focus + .label &,
  .input:not([readonly]):not([disabled]):active + .label & {
    outline: none;
    border-color: transparent;
    .focus();
  }
}

.root {
  display: inline-block;
}

.input {
  opacity: 0;
  position: absolute;
}

.label {
  display: block;
}

.fullWidth {
  width: 100%;
}

.standard {
  .touchableText(@whistling-type-scale-mobile);
  cursor: pointer;
  display: flex;
  align-items: flex-start;
  line-height: @grid-base * 6;
  padding: @grid-base 0;
  @media @desktop {
    .touchableText(@whistling-type-scale);
    line-height: @grid-base * 6;
  }

  .fullWidth & {
    justify-content: space-between;
    width: 100%;
    height: auto;
  }
}

.checkBox {
  .activeFocusReset();

  background-color: @saWhite;
  width: @grid-row-height * 6;
  height: @grid-row-height * 6;
  border: @field-border-width solid @sk-mid-gray-light;
  border-radius: @field-border-radius;
  padding: @field-border-radius;
  fill: @sk-pink;
  align-self: flex-start;

  &.checked {
    background-color: @actionBlue;
    border: @field-border-width solid @actionBlue;
  }
}

.checkBoxLeft {
  margin-right: @field-gutter-width;
}

.checkBoxRight {
  margin-left: @field-gutter-width;
}

.checkMark {
  display: block;
  height: 100%;
  fill: @saWhite;
  opacity: 0;

  .input:checked + .label & {
    opacity: 1;
  }
}

.button {
  .activeFocusReset();
  .touchableText(@whistling-type-scale-mobile);
  display: flex;
  justify-content: center;
  background-color: @saWhite;
  color: @saGrey2;
  border-radius: @field-border-radius;
  cursor: pointer;
  @media @desktop {
    .touchableText(@whistling-type-scale);
  }

  .input:checked + .label & {
    color: @saWhite;
    background-color: @actionBlue;
  }
}
