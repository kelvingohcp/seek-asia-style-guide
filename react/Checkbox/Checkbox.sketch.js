import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from './Checkbox';

const Container = ({ children }) => <div style={{ maxWidth: '400px' }}>{children}</div>;
Container.propTypes = {
  children: PropTypes.node.isRequired
};

const noop = () => {};

export const symbols = {
  'Checkbox/Standard/Unchecked without label': (
    <Container>
      <Checkbox id="checkbox1" onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Checked': (
    <Container>
      <Checkbox id="checkbox2" label="Checkbox" checked={true} onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Unchecked': (
    <Container>
      <Checkbox id="checkbox3" label="Checkbox" onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Long label': (
    <Container>
      <Checkbox id="checkbox4" label="Nullam sodales quis velit nec accumsan. Vivamus quis ex euismod." onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Long Label with sub-label': (
    <Container>
      <Checkbox id="checkbox4" label="Nullam sodales quis velit nec accumsan." extraLabel="Vivamus" onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Checkbox at the end': (
    <Container>
      <Checkbox id="checkbox5" label="Checkbox" rtl={true} onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Uncheck and disabled': (
    <Container>
      <Checkbox id="checkbox6" label="Checkbox" disabled onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Checked and disabled': (
    <Container>
      <Checkbox id="checkbox7" label="Checkbox" checked disabled onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Full width': (
    <Container>
      <Checkbox id="checkbox8" label="Checkbox" fullWidth onChange={noop} />
    </Container>
  ),
  'Checkbox/Standard/Full width and checkbox at the end': (
    <Container>
      <Checkbox id="checkbox9" label="Checkbox" fullWidth rtl={true} onChange={noop} />
    </Container>
  ),
  'Checkbox/Compact/Checked': (
    <Container>
      <Checkbox id="checkbox10" label="Checkbox" compact checked={true} onChange={noop} />
    </Container>
  ),
  'Checkbox/Compact/Unchecked': (
    <Container>
      <Checkbox id="checkbox11" label="Checkbox" compact onChange={noop} />
    </Container>
  )
};

