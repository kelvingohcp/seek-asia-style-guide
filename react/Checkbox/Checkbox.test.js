import React from 'react';
import { shallow } from 'enzyme';
import Checkbox from './Checkbox';

describe('Checkbox', () => {
  const requiredProps = {
    id: 'testCheckbox',
    label: 'Still in role',
    checked: false,
    onChange: () => {}
  };

  it('should render with simple props', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as checked', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} checked={true} />);
    const inputChecked = wrapper.find('input').prop('checked');
    expect(inputChecked).toEqual(true);
  });

  it('should pass through the value', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} value="foo" checked={true} />);
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toEqual('foo');
  });

  it('should render with className', () => {
    const className = 'testClassname';
    const wrapper = shallow(<Checkbox {...requiredProps} className={className} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with standard checkbox style', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} type="standard" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with compact checkbox style', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} type="standard" compact />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with button style', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} type="button" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with checkbox put at the end of component', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} rtl={true} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render full width checkbox', () => {
    const wrapper = shallow(<Checkbox {...requiredProps} fullWidth />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('inputProps', () => {
    it('should invoke the change handler', () => {
      const handleChange = jest.fn();
      const props = { ...requiredProps, inputProps: { onChange: handleChange, checked: false } };
      const checkedEvent = { target: { checked: true } };
      const wrapper = shallow(<Checkbox {...props} />);
      wrapper.find('input').simulate('change', checkedEvent);
      expect(handleChange).toBeCalledWith(checkedEvent);
    });

    it('should render as checked', () => {
      const wrapper = shallow(<Checkbox {...requiredProps} inputProps={{ checked: true }} />);
      const inputChecked = wrapper.find('input').prop('checked');
      expect(inputChecked).toEqual(true);
    });

    it('should pass through other props to the input', () => {
      const wrapper = shallow(<Checkbox {...requiredProps} inputProps={{ checked: false, 'data-automation': 'first-name-field' }} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
