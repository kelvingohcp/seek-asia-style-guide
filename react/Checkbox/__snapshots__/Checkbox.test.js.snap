// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox inputProps should pass through other props to the input 1`] = `
<div
  className="root"
>
  <input
    checked={false}
    className="input"
    data-automation="first-name-field"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={false}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox should render full width checkbox 1`] = `
<div
  className="root fullWidth"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={false}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox should render with button style 1`] = `
<div
  className="root"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label btn"
    htmlFor="testCheckbox"
  >
    <span
      className="mainLabel"
    >
      Still in role
    </span>
  </label>
</div>
`;

exports[`Checkbox should render with checkbox put at the end of component 1`] = `
<div
  className="root"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard rtl"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={false}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox should render with className 1`] = `
<div
  className="root testClassname"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={false}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox should render with compact checkbox style 1`] = `
<div
  className="root compact"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={true}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox should render with simple props 1`] = `
<div
  className="root"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={false}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;

exports[`Checkbox should render with standard checkbox style 1`] = `
<div
  className="root"
>
  <input
    checked={false}
    className="input"
    disabled={false}
    id="testCheckbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="label"
    htmlFor="testCheckbox"
  >
    <div
      className="standard"
    >
      <div
        className="checkBox"
      >
        <Icon
          animation=""
          className=""
          lineHeight=""
          rotation=""
          size=""
          smoothRotate={false}
          svgClassName="checkMark"
          type="check"
        />
      </div>
      <span
        className="mainLabel"
      >
        <Text
          baseline={false}
          className="standardLabel"
          intimate={false}
          strong={false}
        >
          Still in role
        </Text>
      </span>
    </div>
  </label>
</div>
`;
